### PR TITLE
Relaxing divisibility constraints on num_canonical_nodes and num_physical_nodes

### PR DESCRIPTION
**Pull request by snarayan21**
## Description of changes:
Relaxes constraints on `num_canonical_nodes` being divisible by `physical_nodes` or vice versa, opening up many more numbers of nodes to deterministically train and resume on. This can be enabled by setting `partition_algo='reserved'` in StreamingDataset instantiation, or by setting `partition_algo: reserved` in yamls for each dataset. The only assumptions we make are that:
* `num_canonical_nodes` plays nicely with `physical_nodes` only in the initial run.
* global batch size stays constant during resumption.
* global batch size is divisible by the total number of devices during resumption.
This `reserved` partition algo also preserves earlier functionality with determinism and should replace the `orig` algorithm as the default, since `orig` functionality is a subset of this partitioning algo (in an upcoming PR).

Testing:
* Added unit tests proving deterministic sample ordering with multiple numbers of nodes that were not possible with the `orig` partition algorithm
* Conducted local CPU tests with saving and loading dataset state dict to ensure that values were being set and read in correctly
* Conducted resumption tests on interactive instance with 4 GPUs to ensure that state dict was being set and loaded correctly
* Conducted multi-stream elastic deterministic resumption tests going from 2 to 3 nodes with `num_canonical_nodes` set to 2 and 64. This would not have been possible earlier, and results are shown below. Some numerical instability on newer clusters is likely responsible for slight deviations.
* Conducted tests going from 1->2 nodes, to make sure behavior was same as `orig` partitioning.
* Conducted tests going from 2 -> 3 -> 4 nodes, to make sure the initial physical nodes were being persisted in the state and partitions were still deterministic across multiple resumptions.

How it works:
* if partitioning for the first time, use the `orig` partition method. Save the initial number of physical nodes in the state.
* if resuming, partition over the initial number of physical nodes and the initial device batch size, then repartition this across the new number of physical nodes and the new device batch size. `orig` partition is still used if NCN and PN play nicely to preserve all earlier functionality and have nicer downloading for nice numbers of PN.

In the worst case, a node during the resumed run will have to do twice as much shard downloads as a node from the initial run. This is because of the way a global batch has to be split across the new number of nodes. When combined with better shuffling algorithms and batching methods, this should be an acceptable tradeoff.

Test with `num_canonical_nodes=2`.
* Light blue: full 200 step run on 2 nodes.
* Brown: first 100 steps on 2 nodes.
* Gray: last 100 steps on 3 nodes.
Screenshot 2023-10-20 at 2 18 47 AM

Test with `num_canonical_nodes=64`.
* Blue: full 200 step run on 2 nodes.
* Red: first 100 steps on 2 nodes.
* Yellow: last 100 steps on 3 nodes.
Screenshot 2023-10-20 at 2 19 22 AM

Test from 1->2 nodes, preserving `orig` partitioning behavior:
* Blue: full 200 step run on 1 node.
* Red: first 100 steps on 2 nodes.
* Green: last 100 steps on 2 nodes.
Screenshot 2023-10-20 at 12 11 59 PM

Test from 2->3->4 nodes, with NCN=2, to test multiple resumptions:
* Pink: full 200 step run on 2 nodes.
* Purple: first 100 steps on 2 nodes.
* Orange: middle 100 steps on 3 nodes.
* Light Blue: last 100 steps on 4 nodes.
Screenshot 2023-10-20 at 4 29 34 PM

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: streaming/base/dataset.py [hunk 1]
```diff
@@ -678,6 +682,7 @@ def _resume(self, world: World, epoch: int) -> Tuple[int, int]:
         sample_in_epoch = obj['sample_in_epoch']
         self.num_canonical_nodes = obj['num_canonical_nodes']
         self.shuffle_seed = obj['shuffle_seed']
+        self.initial_physical_nodes = obj['initial_physical_nodes']
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply streaming/base/dataset.py:685]: ooh nice catch. yeah I'll default it to None if it's not there.

File: streaming/base/dataset.py [hunk 2]
```diff
@@ -341,6 +341,10 @@ def __init__(self,
         self.shuffle_block_size = shuffle_block_size
         self.batching_method = batching_method
 
+        # Initialize initial_physical_nodes to None. If we are resuming, then we will set it to the
+        # number of physical nodes of the initial run in the _resume function.
+        self.initial_physical_nodes = None
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply streaming/base/dataset.py:346]: Just addressed a potential bug that would have occurred with multiple resumptions. I'm also going to test that out now. Thanks for pointing this out!!

File: streaming/base/partition/relaxed.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            partitioned over the workers. Defaults to ``None``.
+        drop_first (int): Number of samples seen already, which are dropped. Defaults to ``0``.
+        initial_physical_nodes (int, optional): Number of physical nodes at the start of training.
+            Defaults to ``None``.
+
+    Returns:
+        NDArray[np.int64]: Partitions of shape (physical nodes, ranks per node, workers per rank,
+            batches per worker, batch size).
+    """
+    if num_samples <= drop_first:
```
[the remark this answers is not in the document]
Comment (snarayan21) [reply streaming/base/partition/relaxed.py:52]: This was already in `get_partitions_orig()`, I just copied it over here too :)

File: streaming/base/dataset.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (snarayan21) [reply streaming/base/dataset.py:346]: Added a test (last graph in the PR description going from 2->3->4 nodes with NCN=2 to make sure this all works with multiple elastic resumptions!

File: streaming/base/dataset.py [hunk 4]
```diff
@@ -678,6 +682,11 @@ def _resume(self, world: World, epoch: int) -> Tuple[int, int]:
         sample_in_epoch = obj['sample_in_epoch']
         self.num_canonical_nodes = obj['num_canonical_nodes']
         self.shuffle_seed = obj['shuffle_seed']
+        if 'initial_physical_nodes' in obj:
```
Comment (karan6181) [inline streaming/base/dataset.py:685]: nit: Does this `obj.get('initial_physical_nodes', None)` works?

File: streaming/base/dataset.py [hunk 5]
```diff
@@ -732,11 +741,21 @@ def state_dict(self, num_samples: int, from_beginning: bool) -> Dict[str, Any]:
             sample_in_epoch = num_samples
         else:
             sample_in_epoch = offset + num_samples
+
+        if self.initial_physical_nodes is None:
+            # In this case, we are running for the firs time, so we set initial_physical_nodes
```
Comment (karan6181) [inline streaming/base/dataset.py:746]:
```suggestion
            # In this case, we are running for the first time, so we set initial_physical_nodes
```

File: tests/test_partition.py [hunk 6]
```diff
@@ -1,11 +1,13 @@
 # Copyright 2023 MosaicML Streaming authors
 # SPDX-License-Identifier: Apache-2.0
 
+import pytest
```
Comment (karan6181) [inline tests/test_partition.py:4]: Can we also have a test that validates that no sample duplication and no drop samples when doing a resumption? You might have to track the sample ID and check if there is any sample duplication or drop samples!

File: streaming/base/dataset.py (hunk 4 above)
Comment (snarayan21) [reply streaming/base/dataset.py:685]: yes totally. changed it

File: streaming/base/dataset.py (hunk 5 above)
Comment (snarayan21) [reply streaming/base/dataset.py:746]: changed!

File: tests/test_partition.py (hunk 6 above)
Comment (snarayan21) [reply tests/test_partition.py:4]: I'll modify the test to do the following:
* If the NCN and PN value was supported by the `orig` partitioning algo, then make sure the output of `relaxed` partition is the same as the `orig` partition.
* If the NCN and PN value was not supported by the `orig` partitioning algo, then make sure that the output of the `relaxed` partition, when split up by global batches, is **the exact same** as the initial sample partition. `relaxed` partitioning is written so that this condition must be true. This means that there is no additional sample duplication and no additional dropped samples compared to the initial sample partition when doing a resumption.

File: streaming/base/partition/relaxed.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+import numpy as np
+from numpy.typing import NDArray
+
+from streaming.base.partition.orig import get_partitions_orig
+
+logger = logging.getLogger(__name__)
+
+
+def get_partitions_relaxed(num_samples: int,
```
Comment (karan6181) [inline streaming/base/partition/relaxed.py:17]: Can we add some `logger.debug()` messages in this method and other places that display the shapes or any additional helpful information when a user enables the logger to debug mode?

Comment (karan6181) [reply streaming/base/partition/relaxed.py:17]: Based on the offline discussion, let's track adding a logger.debug messages in this repo as part of another PR.

**Review by karan6181 — APPROVED** [review]
LGTM. Thank You!
